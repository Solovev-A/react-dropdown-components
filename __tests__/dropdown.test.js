import React from 'react';
import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom';
import 'jest-styled-components';
import userEvent from "@testing-library/user-event";

import Dropdown from './../lib/components/dropdown';

const requiredProps = {
    parentBounds: { width: 5, height: 5 }
};

const options = ['Apple', 'Google'];

describe('Dropdown component', () => {
    describe('renders correctly', () => {
        test('with required props only', () => {
            const { container } = render(<Dropdown {...requiredProps} />);
            expect(container.firstChild).toMatchSnapshot();
        });

        test('at top position', () => {
            const { container } = render(<Dropdown {...requiredProps} position="top" />);
            expect(container.firstChild).toMatchSnapshot();
        });

        test('at bottom position', () => {
            const { container } = render(<Dropdown {...requiredProps} position="bottom" />);
            expect(container.firstChild).toMatchSnapshot();
        });
    });

    it('use maxHeight prop', () => {
        const { container } = render(<Dropdown {...requiredProps} maxHeight={500} />);
        expect(container.firstChild).toHaveStyleRule('max-height', '500px');
    });

    it('width equal parent\'s width', () => {
        const { container } = render(<Dropdown {...requiredProps} />);
        expect(container.firstChild).toHaveStyleRule('width', `${requiredProps.parentBounds.width}px`);
    });

    it('renders content instead of options', () => {
        const props = {
            content: <span>Test</span>,
            options
        };
        render(<Dropdown {...requiredProps} {...props} />);
        expect(screen.getByText('Test')).toBeInTheDocument();
        expect(screen.queryAllByRole('option').length).toBe(0);
    });

    it('renders options', () => {
        render(<Dropdown {...requiredProps} options={options} />);
        expect(screen.getByText(/google/i)).toBeInTheDocument();
        expect(screen.getAllByRole('option').length).toBe(2);
    });

    describe('behavior', () => {
        it('use callback on select option', () => {
            const onSelectOption = jest.fn();
            render(<Dropdown {...requiredProps} options={options} onSelectOption={onSelectOption} />);

            fireEvent(screen.getAllByRole('option')[0], new MouseEvent('click', {
                bubbles: true,
                cancelable: true,
            }));

            expect(onSelectOption).toHaveBeenCalledTimes(1);
            expect(onSelectOption).toHaveBeenCalledWith('Apple');
        });

        it('use callback on update pointer', () => {
            const onUpdatePointer = jest.fn();
            render(<Dropdown {...requiredProps} options={options} onUpdatePointer={onUpdatePointer} />);

            const optionElements = screen.getAllByRole('option');

            userEvent.hover(optionElements[1]);

            expect(onUpdatePointer).toHaveBeenCalledTimes(1);
            expect(onUpdatePointer).toHaveBeenCalledWith(1);

            userEvent.hover(optionElements[0]);

            expect(onUpdatePointer).toHaveBeenCalledTimes(2);
            expect(onUpdatePointer).toHaveBeenCalledWith(0);
        });
    });
});