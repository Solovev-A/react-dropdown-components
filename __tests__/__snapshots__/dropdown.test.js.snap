// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component renders correctly at bottom position 1`] = `
.c1 {
  padding: .375rem .75rem;
  display: block;
  white-space: nowrap;
  cursor: pointer;
}

.c2 {
  cursor: default;
}

.c0 {
  position: absolute;
  left: -1px;
  padding: 0px;
  max-height: 350px;
  width: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-x: hidden;
  border-radius: 0;
  background-color: white;
  border: 1px solid #ced4da;
  box-sizing: border-box;
  z-index: 19;
  top: 4px;
  border-top: none;
  border-bottom-right-radius: .25rem;
  border-bottom-left-radius: .25rem;
}

<div
  class="c0"
>
  <span
    class="c1 c2"
  >
    Совпадений не найдено
  </span>
</div>
`;

exports[`Dropdown component renders correctly at top position 1`] = `
.c1 {
  padding: .375rem .75rem;
  display: block;
  white-space: nowrap;
  cursor: pointer;
}

.c2 {
  cursor: default;
}

.c0 {
  position: absolute;
  left: -1px;
  padding: 0px;
  max-height: 350px;
  width: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-x: hidden;
  border-radius: 0;
  background-color: white;
  border: 1px solid #ced4da;
  box-sizing: border-box;
  z-index: 19;
  bottom: 5px;
  border-bottom: none;
  border-top-right-radius: .25rem;
  border-top-left-radius: .25rem;
}

<div
  class="c0"
>
  <span
    class="c1 c2"
  >
    Совпадений не найдено
  </span>
</div>
`;

exports[`Dropdown component renders correctly with required props only 1`] = `
.c1 {
  padding: .375rem .75rem;
  display: block;
  white-space: nowrap;
  cursor: pointer;
}

.c2 {
  cursor: default;
}

.c0 {
  position: absolute;
  left: -1px;
  padding: 0px;
  max-height: 350px;
  width: 5px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow-x: hidden;
  border-radius: 0;
  background-color: white;
  border: 1px solid #ced4da;
  box-sizing: border-box;
  z-index: 19;
  top: 4px;
  border-top: none;
  border-bottom-right-radius: .25rem;
  border-bottom-left-radius: .25rem;
}

<div
  class="c0"
>
  <span
    class="c1 c2"
  >
    Совпадений не найдено
  </span>
</div>
`;
